concordium-consensus/src/Concordium/KonsensusV1/Scheduler.hs

```diff
@@ -1,45 +1,200 @@
+--      - Updates the time and mint rate for the next payday, based on the current chain parameters.
+--
+--      - Process pending cooldowns on bakers and delegators that were set to elapse by the
+--        trigger block time for the previous epoch.
 --
 --  * The seed state is updated to reflect the epoch transition.
 --
--- Note: this does not update the next epoch bakers or capital distribution, which should be done
--- subsequently where necessary. This also does not update the seed state to account for the block
--- nonce, which should also be done subsequently.
+--  * Of the new epoch is the epoch before the next payday, take a snapshot of bakers and
+--    delegators, allowing for cooldowns that are set to elapse at by the trigger block time for
+--    this epoch.
+--
+-- Note: If the baker or delegator cooldown period is ever less than the duration of an epoch, then
+-- it would be possible to have a baker not in cooldown when the baker snapshot is taken, but be
+-- removed when the cooldowns are processed at the payday. This bad, because the baker/delegator
+-- would not have their stake locked while they are baking/delegating. However, this should not be
+-- a catastrophic invariant violation.
 doEpochTransition ::
+    forall m.
     (BlockStateOperations m, IsConsensusV1 (MPV m)) =>
     -- |Whether the block is the first in a new epoch
     Bool ->
     -- |The epoch duration
     Duration ->
     -- |State to update
     UpdatableBlockState m ->
-    m (UpdatableBlockState m)
-doEpochTransition False _ theState = return theState
+    m (Maybe PaydayParameters, UpdatableBlockState m)
+doEpochTransition False _ theState = return (Nothing, theState)
 doEpochTransition True epochDuration theState = do
-    theState <- bsoRotateCurrentCapitalDistribution theState
-    theState <- bsoRotateCurrentEpochBakers theState
+    chainParams <- bsoGetChainParameters theState
     oldSeedState <- bsoGetSeedState theState
+    let newEpoch = (oldSeedState ^. epoch) + 1
+    nextPayday <- bsoGetPaydayEpoch theState
+    (theState, mPaydayParams, newNextPayday) <-
+        if newEpoch == nextPayday
+            then do
+                -- We grab the current payday parameters to use later for minting and distributing
+                -- rewards, because we will overwrite them with the parameters for the new payday.
+                paydayCapitalDistribution <- bsoGetCurrentCapitalDistribution theState
+                paydayBakers <- bsoGetCurrentEpochFullBakersEx theState
+                paydayPoolRewards <- bsoGetBakerPoolRewardDetails theState
+                paydayMintRate <- bsoGetPaydayMintRate theState
+                let paydayParams = PaydayParameters{..}
+                -- Rotate the capital distribution and bakers so that the current values are
+                -- replaced by the next values.
+                theState <- bsoRotateCurrentCapitalDistribution theState
+                theState <- bsoRotateCurrentEpochBakers theState
```

Comment:
I think it is a really bad idea to reuse variable names in Haskell. We don't do this anywhere else so please don't do it here either. 

It's so easy to get wrong.

Reply:
I will make the change, but I disagree. Not reusing names in this case makes it much easier to get things wrong.